File: Music.Android/short/app/src/main/java/com/zitherharpmusic/zhmshort/ui/artist/ArtistListAdapter.java
package com.zitherharpmusic.zhmshort.ui.artist;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.zitherharpmusic.zhmshort.R;
import com.zitherharpmusic.zhmshort.model.Artist;
import com.zitherharpmusic.zhmshort.model.Language;
import com.zitherharpmusic.zhmshort.model.PhotoQuality;
import com.zitherharpmusic.zhmshort.util.ListenerUtils;
import com.zitherharpmusic.zhmshort.util.MainUtils;

import java.util.List;

public class ArtistListAdapter extends RecyclerView.Adapter<ArtistListAdapter.ViewHolder> {
    private final FragmentActivity fragmentActivity;
    private final List<Artist> artists;

    public ArtistListAdapter(FragmentActivity fragmentActivity, List<Artist> artists) {
        this.fragmentActivity = fragmentActivity;
        this.artists = artists;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_artist_list, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Artist artist = artists.get(position);
        holder.title.setText(artist.getName(Language.VIETNAMESE));
        holder.subtitle.setText(artist.getName(Language.SIMPLIFIED_CHINESE));
        holder.itemView.setOnClickListener(ListenerUtils.launchActivity(fragmentActivity, ArtistActivity.class, artist));
        MainUtils.loadImage(fragmentActivity, holder.photo, artist.getPhotoUrl(PhotoQuality.SMALL));
    }

    @Override
    public int getItemCount() {
        return artists.size();
    }

    protected static class ViewHolder extends RecyclerView.ViewHolder {
        protected ImageView photo;
        protected TextView title;
        protected TextView subtitle;

        protected ViewHolder(View view) {
            super(view);
            photo = view.findViewById(R.id.photo);
            title = view.findViewById(R.id.title);
            subtitle = view.findViewById(R.id.subtitle);
        }
    }
}

File: Music.Client.Android/Kotlin/Music/short/src/main/java/com/zitherharp/music/shorts/ui/artist/ArtistListAdapter.kt
package com.zitherharp.music.shorts.ui.artist

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.zitherharp.music.Extension.setImageUrl
import com.zitherharp.music.Language
import com.zitherharp.music.core.QQMusic
import com.zitherharp.music.model.Artist
import com.zitherharp.music.shorts.databinding.ArtistListContentBinding
import com.zitherharp.music.ui.RecyclerViewAdapter

class ArtistListAdapter(private val context: Context,
                        private val artists: List<Artist>):
    RecyclerViewAdapter<ArtistListAdapter.ViewHolder>(context, artists) {

    inner class ViewHolder(binding: ArtistListContentBinding):
        RecyclerView.ViewHolder(binding.root) {
        val artistImage = binding.artistImage
        val artistChineseName = binding.artistChineseName
        val artistVietnameseName = binding.artistVietnameseName
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ViewHolder(
            ArtistListContentBinding.inflate(
                LayoutInflater.from(parent.context), parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        with(holder) {
            with(artists[position]) {
                itemView.setOnClickListener {
                    context.startActivity(Intent(context, ArtistDetailActivity::class.java).apply {
                        putExtra(ArtistDetailActivity::class.simpleName, id)
                    })
                }
                artistImage.setImageUrl(getImageUrl(QQMusic.Image.SMALL))
                artistChineseName.text = getName(Language.CHINESE)
                artistVietnameseName.text = getName(Language.VIETNAMESE)
            }
        }
    }
}